pyro/distributions/hmm.py

```diff
@@ -0,0 +1,95 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints
+
+from pyro.distributions.torch_distribution import TorchDistribution
+from pyro.distributions.util import broadcast_shape
+
+
+def _logmatmulexp(x, y):
+    """
+    Numerically stable version of ``(x.log() @ y.log()).exp()``.
+    """
+    x_shift = x.max(-1, keepdim=True)[0]
+    y_shift = y.max(-2, keepdim=True)[0]
+    xy = torch.matmul((x - x_shift).exp(), (y - y_shift).exp()).log()
+    return xy + x_shift + y_shift
```

Comment:
Nice trick! I think you can also use [this implementation](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/transforms/block_autoregressive.py#L18) for simplicity.

Reply:
`torch.matmul()` is faster and uses less memory than broadcast-and-contract.

Reply:
That makes sense, thanks for explaining!